packages/tldraw/src/lib/shapes/shared/useEmojis.ts

```diff
@@ -0,0 +1,125 @@
+import { useState } from 'react'
+import { useDialogs } from '../../ui/hooks/useDialogsProvider'
+import EmojiDialog from './emojis'
+import { EmojiDialogSingleton } from './emojis/EmojiDialog'
+
+export function useEmojis(inputEl: HTMLTextAreaElement | null, onComplete: (text: string) => void) {
+	const { addDialog, removeDialog } = useDialogs()
+	const [emojiSearchText, setEmojiSearchText] = useState('')
+	const [isEmojiMenuOpen, setIsEmojiMenuOpen] = useState(false)
+
+	const closeMenu = () => {
+		setIsEmojiMenuOpen(false)
+		removeDialog('emoji')
+		setEmojiSearchText('')
+	}
+
+	const onEmojiSelect = (emoji: any) => {
+		if (!inputEl) return
+
+		const searchText = EmojiDialogSingleton?.component.refs.searchInput.current.value
+		inputEl.focus()
+		inputEl.setSelectionRange(
+			inputEl.selectionStart - searchText.length - 1,
+			inputEl.selectionStart
+		)
+		inputEl.setRangeText(emoji.native)
+		inputEl.setSelectionRange(inputEl.selectionStart + 1, inputEl.selectionStart + 1)
```

Comment:
you could `inputEl.setRangeText(emoji.native, undefined, undefined, 'end')` if u wanted

Reply:
hm, i see what you're going for but that overloaded sig. doesn't allow undefined and even if i rework it with the correct values, i think splitting it out is a little clearer on the intent.